File: packages/decart_realtime_bridge/android/src/main/kotlin/com/morphly/decart_realtime_bridge/DecartRealtimeBridgePlugin.kt
package com.morphly.decart_realtime_bridge

import android.os.Handler
import android.os.Looper
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class DecartRealtimeBridgePlugin : FlutterPlugin, MethodChannel.MethodCallHandler, EventChannel.StreamHandler {
    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var eventSink: EventChannel.EventSink? = null
    private val handler = Handler(Looper.getMainLooper())
    private var elapsedSeconds = 0
    private var ticking = false

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel = MethodChannel(binding.binaryMessenger, "morphly/decart_realtime/methods")
        eventChannel = EventChannel(binding.binaryMessenger, "morphly/decart_realtime/events")
        methodChannel.setMethodCallHandler(this)
        eventChannel.setStreamHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        stopTicker("stopped", "Bridge detached.")
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "startSession" -> {
                val clientToken = call.argument<String>("clientToken")
                val model = call.argument<String>("model")
                val prompt = call.argument<String>("prompt")
                if (clientToken.isNullOrBlank() || model.isNullOrBlank()) {
                    emit("failed", "Missing Decart token or model.", elapsedSeconds)
                    result.error("invalid_args", "Missing Decart token or model.", null)
                    return
                }

                // Production hook:
                // 1. Initialize ai.decart.sdk.DecartClient with the short-lived token.
                // 2. Attach the camera WebRTC track using the selected quality/model constraints.
                // 3. Send the reference image/prompt and render remote frames into a PlatformView.
                emit("starting", "Connecting to Decart.", 0)
                startTicker(prompt ?: "Morphly realtime session")
                result.success(null)
            }
            "setPrompt" -> {
                val prompt = call.argument<String>("prompt") ?: ""
                // Production hook: client.realtime.setPrompt(prompt, enhance = true)
                emit("connected", "Prompt updated: $prompt", elapsedSeconds)
                result.success(null)
            }
            "stopSession" -> {
                stopTicker("stopped", "Session stopped.")
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
        emit("idle", "Ready.", 0)
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    private fun startTicker(message: String) {
        stopTicker("starting", message)
        elapsedSeconds = 0
        ticking = true
        emit("connected", message, elapsedSeconds)
        handler.postDelayed(object : Runnable {
            override fun run() {
                if (!ticking) return
                elapsedSeconds += 1
                emit("connected", "Live morphing.", elapsedSeconds)
                handler.postDelayed(this, 1000)
            }
        }, 1000)
    }

    private fun stopTicker(state: String, message: String) {
        ticking = false
        handler.removeCallbacksAndMessages(null)
        emit(state, message, elapsedSeconds)
    }

    private fun emit(state: String, message: String, elapsed: Int) {
        handler.post {
            eventSink?.success(mapOf(
                "state" to state,
                "message" to message,
                "elapsedSeconds" to elapsed
            ))
        }
    }
}

File: packages/decart_realtime_bridge/ios/Classes/DecartRealtimeBridgePlugin.swift
import Flutter
import UIKit

public class DecartRealtimeBridgePlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
  private var eventSink: FlutterEventSink?
  private var timer: Timer?
  private var elapsedSeconds = 0

  public static func register(with registrar: FlutterPluginRegistrar) {
    let methods = FlutterMethodChannel(
      name: "morphly/decart_realtime/methods",
      binaryMessenger: registrar.messenger()
    )
    let events = FlutterEventChannel(
      name: "morphly/decart_realtime/events",
      binaryMessenger: registrar.messenger()
    )
    let instance = DecartRealtimeBridgePlugin()
    registrar.addMethodCallDelegate(instance, channel: methods)
    events.setStreamHandler(instance)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "startSession":
      guard
        let args = call.arguments as? [String: Any],
        let token = args["clientToken"] as? String,
        let model = args["model"] as? String,
        !token.isEmpty,
        !model.isEmpty
      else {
        emit(state: "failed", message: "Missing Decart token or model.", elapsed: elapsedSeconds)
        result(FlutterError(code: "invalid_args", message: "Missing Decart token or model.", details: nil))
        return
      }

      // Production hook:
      // 1. Add https://github.com/decartai/decart-ios.git in Xcode Swift Package Manager.
      // 2. Initialize DecartClient with this short-lived token.
      // 3. Connect the camera track, reference image, and prompt to Lucy realtime.
      // 4. Render the remote stream through a Flutter PlatformView.
      emit(state: "starting", message: "Connecting to Decart.", elapsed: 0)
      startTicker(message: (args["prompt"] as? String) ?? "Morphly realtime session")
      result(nil)
    case "setPrompt":
      let args = call.arguments as? [String: Any]
      let prompt = args?["prompt"] as? String ?? ""
      emit(state: "connected", message: "Prompt updated: \(prompt)", elapsed: elapsedSeconds)
      result(nil)
    case "stopSession":
      stopTicker(state: "stopped", message: "Session stopped.")
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    eventSink = events
    emit(state: "idle", message: "Ready.", elapsed: 0)
    return nil
  }

  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    eventSink = nil
    return nil
  }

  private func startTicker(message: String) {
    stopTicker(state: "starting", message: message)
    elapsedSeconds = 0
    emit(state: "connected", message: message, elapsed: elapsedSeconds)
    timer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { [weak self] _ in
      guard let self else { return }
      self.elapsedSeconds += 1
      self.emit(state: "connected", message: "Live morphing.", elapsed: self.elapsedSeconds)
    }
  }

  private func stopTicker(state: String, message: String) {
    timer?.invalidate()
    timer = nil
    emit(state: state, message: message, elapsed: elapsedSeconds)
  }

  private func emit(state: String, message: String, elapsed: Int) {
    eventSink?([
      "state": state,
      "message": message,
      "elapsedSeconds": elapsed
    ])
  }
}
